### PR TITLE
Productize QuickGrid

### LINKED ISSUE #46317: Productize QuickGrid
For .NET 7, QuickGrid was experimental and still lives in a "preview"-branded package.

For .NET 8, can we make it a built-in supported part of the framework? Note that there are [various feature requests](https://github.com/dotnet/aspnetcore/issues?q=is%3Aissue+is%3Aopen+label%3Afeature-blazor-quickgrid) we should consider as part of this effort.

### LINKED ISSUE #46789: [API Proposal] Productize QuickGrid
## Background and Motivation

The [QuickGrid package](https://github.com/aspnet/AspLabs/tree/main/src/QuickGrid) aims to provide a simple and performant data grid component for Blazor. QuickGrid provides some basic functionality such as sorting, paging, styling and virtualization but aims to provide extensibility to developers that want to build on it.

Even through productization, the main goals of QuickGrid are:

- To provide a reference architecture for a performant data grid component, which others may build on or copy.
- To provide a convenient and reliable standalone free data grid component to Blazor developers who don't require advanced grid features.

QuickGrid can show data from three main sources:

- In-memory IQueryable
- EF Core IQueryable
- Remote data

CC: @SteveSandersonMS

### The following propose new API in 2 new namespaces:

```cs
namespace Microsoft.AspNetCore.Components.QuickGrid;

public partial class QuickGrid<TGridItem> : IAsyncDisposable
{
    public QuickGrid();
    public IQueryable<TGridItem>? Items { get; set; }
    public GridItemsProvider<TGridItem>? ItemsProvider { get; set; }
    public string? Class { get; set; }
    public string? Theme { get; set; }
    public RenderFragment? ChildContent { get; set; }
    public bool Virtualize { get; set; }
    public float ItemSize { get; set; }
    public bool ResizableColumns { get; set; }
    public Func<TGridItem, object> ItemKey { get; set; }
    public PaginationState? Pagination { get; set; }
    protected override Task OnParametersSetAsync();
    protected override async Task OnAfterRenderAsync(bool firstRender);
    public Task SortByColumnAsync(ColumnBase<TGridItem> column, SortDirection direction);
    public void ShowColumnOptions(ColumnBase<TGridItem> column);
    public async Task RefreshDataAsync();
    public async ValueTask DisposeAsync();
}

public enum Align
{
    Left,
    Center,
    Right,
}

public abstract partial class ColumnBase<TGridItem>
{
    public ColumnBase();
    public string? Title { get; set; }
    public string? Class { get; set; }
    public Align Align { get; set; }
    public RenderFragment<ColumnBase<TGridItem>>? HeaderTemplate { get; set; }
    public RenderFragment? ColumnOptions { get; set; }
    public bool? Sortable { get; set; }
    public SortDirection? IsDefaultSort { get; set; }
    public RenderFragment<PlaceholderContext>? PlaceholderTemplate { get; set; }
    public QuickGrid<TGridItem> Grid;
    protected internal abstract void CellContent(RenderTreeBuilder builder, TGridItem item);
    protected internal RenderFragment HeaderContent { get; protected set; }
    protected virtual bool IsSortableByDefault();
    protected override void BuildRenderTree(RenderTreeBuilder builder);
}

public delegate ValueTask<GridItemsProviderResult<TGridItem>> GridItemsProvider<TGridItem>(
    GridItemsProviderRequest<TGridItem> request);

public readonly struct GridItemsProviderRequest<TGridItem>
{
    public int StartIndex { get; }
    public int? Count { get; }
    public ColumnBase<TGridItem>? SortByColumn { get; }
    public bool SortByAscending { get; }
    public CancellationToken CancellationToken { get; }
    public IQueryable<TGridItem> ApplySorting(IQueryable<TGridItem> source);
    public IReadOnlyCollection<(string PropertyName, SortDirection Direction)> GetSortByProperties();
}

public struct GridItemsProviderResult<TGridItem>
{
    public GridItemsProviderResult(ICollection<TGridItem> items, int totalItemCount);
    public ICollection<TGridItem> Items { get; set; }
    public int TotalItemCount { get; set; }
}

public static class GridItemsProviderResult
{
    public static GridItemsProviderResult<TGridItem> From<TGridItem>(ICollection<TGridItem> items, int totalItemCount);
}

public sealed class GridSort<TGridItem>
{
    public static GridSort<TGridItem> ByAscending<U>(Expression<Func<TGridItem, U>> expression);
    public static GridSort<TGridItem> ByDescending<U>(Expression<Func<TGridItem, U>> expression);
    public static GridSort<TGridItem> ThenAscending<U>(Expression<Func<TGridItem, U>> expression);
    public static GridSort<TGridItem> ThenDescending<U>(Expression<Func<TGridItem, U>> expression);
}

public interface ISortBuilderColumn<TGridItem>
{
    public GridSort<TGridItem>? SortBuilder { get; }
}

public class PaginationState
{
    public int ItemsPerPage { get; set; }
    public int CurrentPageIndex { get; }
    public int? TotalItemCount { get; }
    public int? LastPageIndex;
    public event EventHandler<int?>? TotalItemCountChanged;
    public override int GetHashCode();
    public Task SetCurrentPageIndexAsync(int pageIndex);
}

public partial class Paginator : IDisposable
{
    public Paginator();
    public PaginationState Value { get; set; }
    public RenderFragment? SummaryTemplate { get; set; }
    public void Dispose();
    protected override void OnParametersSet();
    protected override void BuildRenderTree(RenderTreeBuilder builder);
}

public class PropertyColumn<TGridItem, TProp> : ColumnBase<TGridItem>, ISortBuilderColumn<TGridItem>
{
    public Expression<Func<TGridItem, TProp>> Property { get; set; }
    public string? Format { get; set; }
    GridSort<TGridItem>? ISortBuilderColumn<TGridItem>.SortBuilder;
    protected override void OnParametersSet();
    protected internal override void CellContent(RenderTreeBuilder builder, TGridItem item);
}

public enum SortDirection
{
    Ascending,
    Descending,
    Auto,
}

public class TemplateColumn<TGridItem> : ColumnBase<TGridItem>, ISortBuilderColumn<TGridItem>
{
    public RenderFragment<TGridItem> ChildContent { get; set; }
    public GridSort<TGridItem>? SortBy { get; set; }
    GridSort<TGridItem>? ISortBuilderColumn<TGridItem>.SortBuilder;
    protected internal override void CellContent(RenderTreeBuilder builder, TGridItem item);
    protected override bool IsSortableByDefault();
}
```

```cs
namespace Microsoft.AspNetCore.Components.QuickGrid.Infrastructure;

public sealed class ColumnsCollectedNotifier<TGridItem> : IComponent
{
    public void Attach(RenderHandle renderHandle);
    public Task SetParametersAsync(ParameterView parameters);
}

public sealed class Defer : ComponentBase
{
    public RenderFragment? ChildContent { get; set; }
}

public static class EventHandlers
{
}

public interface IAsyncQueryExecutor
{
    bool IsSupported<T>(IQueryable<T> queryable);
    Task<int> CountAsync<T>(IQueryable<T> queryable);
    Task<T[]> ToArrayAsync<T>(IQueryable<T> queryable);
}
```

### New API in an existing namespace:

```diff
namespace Microsoft.Extensions.DependencyInjection;

+public static class EntityFrameworkAdapterServiceCollectionExtensions
+{
+    public static void AddQuickGridEntityFrameworkAdapter(this IServiceCollection services);
+}
```

## Usage Examples

For a complete set of interactive examples, please see https://aspnet.github.io/quickgridsamples/sample.

## Risks

This was originally an experimental package, and the only changes since the first draft of this proposal were to make it compliant with the aspnetcore repo's analyzer/build configuration. This issue will be marked with the appropriate label when ready, but in the meantime feel free to review the attached PR if you are able.

The only other risk I can think of is the cost of additional E2E tests on our CI. The testing story for this package should be designed with that consideration.

### PR DESCRIPTION
**Pull request by Nick-Stanton**
# Productize QuickGrid

Productizes the experimental [QuickGrid package](https://github.com/aspnet/AspLabs/tree/main/src/QuickGrid), which aims to provide a simple and performant data grid component for Blazor.

## Description

There is a [list of issues](https://github.com/dotnet/aspnetcore/issues?q=is%3Aissue+is%3Aopen+label%3Afeature-blazor-quickgrid) that we could address either as part of this PR or include small fixes and additions following a merge here. There is already a lot of new code and API being proposed so non-bug fixes may be best for after this PR.

Following initial review here I will open an API proposal for these changes.

Tested:
- [x] Manually
- [x] E2E

Fixes #46317

### CODE REVIEW

File: src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs
Comment (Nick-Stanton) [inline src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs:37]: The trimmer isn't a fan of this line because `IsEntityFrameworkProviderType` is accessed via reflection. @SteveSandersonMS what do you think about this?

**Comment by dougbu** [comment]
@Nick-Stanton, you might want to instead use the https://github.com/dotnet/aspnetcore-internal/tree/main/scripts/consolidation infrastruture to copy over every relevant file, commit, issue reference, and PR reference from the AspLabs repo. @wtgodbe or I could help you get started on that. Ignore this if the history isn't relevant from the @dotnet/aspnet-blazor-eng perspective.

**Comment by Nick-Stanton** [comment]
> @Nick-Stanton, you might want to instead use the https://github.com/dotnet/aspnetcore-internal/tree/main/scripts/consolidation infrastruture to copy over every relevant file, commit, issue reference, and PR reference from the AspLabs repo. @wtgodbe or I could help you get started on that. Ignore this if the history isn't relevant from the @dotnet/aspnet-blazor-eng perspective.

Thanks @dougbu, TIL. Unless the Blazor team disagrees, I will keep the PR as-is and include a co-author credit to @SteveSandersonMS in the squashed merge commit. We were already tracking issues here so I don't expect a need for the consolidation infra.

File: src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs
Comment (SteveSandersonMS) [reply src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs]: I know it's been a while since you raised this point so am not sure if any clarity has emerged since. Is this still an open question? TBH I don't know for sure what the right thing to do is, but would start by investigating if there are any linker annotations that match the scenario.

**Comment by SteveSandersonMS** [comment]
This looks great to me.

However since I wrote the QuickGrid implementation I'm not sure we get the most value if I'm the one approving this PR :) @Nick-Stanton, if you feel the implementation is clear to you then we're probably good to go with it. But if there are aspects that seem unclear we might also want to pull in at least one other team member to inspect it, share out an understanding of what's going on internally, and maybe spot any opportunities to improve it further.

File: src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs
Comment (Nick-Stanton) [reply src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs]: I'll investigate. In the meantime, would you be willing to clarify the goals of this class? The comments are helpful but I admittedly don't see the full picture. If the type of your `IQueryable<T>` is supported by EF (or comes from EF?) and you're not using the adapter then an error is thrown?

**Review by MackinnonBuck — COMMENTED** [review]
This looks good to me!

Should we consider sealing some (most?) of these public types? I don't see why someone would need to derive from `GridSort<TGridItem>`, or `Defer`, for example.

File: src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs
Comment (SteveSandersonMS) [reply src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs]: The goal is to be able to resolve queries asynchronously when possible. EF gives the developer an `IQueryable<T>` but they also give a set of async extension methods like `myQueryable.CountAsync()`. Those extension methods don't exist if you're not referencing EF and only work on EF-supplied IQueryables.

Since QuickGrid itself doesn't want5a direct dependency on EF (because in most cases it's not wanted in WebAssembly apps, and the linker wouldn't even be able to remove it if we're doing runtime checks to find out whether a given IQueryable comes from EF), the `Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter` package is an optional dependency that registers a DI service that does the runtime check ("is this IQueryable from EF?") and resolves QuickGrid's queries using the EF async extension methods instead of the synchronous IQueryable methods.

The particular line of code here (line 37) is where we've already checked if you have the optional adapter service and found that you do **not** (that's on lines 30-32). So the remaining purpose of this branch is to help the developer notice if they *should* have added the adapter service but forgot to do so. We are using reflection to check whether the current IQueryable is an EF-enabled one, because if it is, the developer has made a mistake in failing to register the adapter service.

As per the comment on lines 54-55, it doesn't matter if this check gets trimmed out in production or if it fails to notice EF-provided IQueryables, because it's only to help detect misconfiguration and is sufficient to do this in development only.

Comment (Nick-Stanton) [reply src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs]: Sounds good. In this case, it looks like we can suppress this warning. The runtime repo has tons of regression testing to ensure that what we'd want doesn't get trimmed out, but this doesn't seem like the time to look into setting it up in aspnetcore. Besides, the specific error (IL2111) is concerned with not guaranteeing meeting the `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.Interfaces)]` annotation. We're okay with the prospect of an interface potentially being trimmed away because as you pointed out, this error is designed to be caught during development.

**Comment by Nick-Stanton** [comment]
> This looks good to me!
> 
> Should we consider sealing some (most?) of these public types? I don't see why someone would need to derive from `GridSort<TGridItem>`, or `Defer`, for example.

Great point. I sealed what I knew for sure would be good, and added `[EditorBrowsable(EditorBrowsableState.Never)]` on a couple infra classes that ideally would be private/internal but must be public. There are public classes without inheritors that I didn't seal, as I could see places that a developer may want to inherit and build on.

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Infrastructure/AsyncQueryExecutorSupplier.cs
Comment (halter73) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Infrastructure/AsyncQueryExecutorSupplier.cs:28]:
```suggestion
    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2111",
               Justification = "The reflection is a best effort to warn developers about sync-over-async behavior which can cause thread pool starvation.")]
```

I agree the trim-unsafe reflection is worth doing because it's better to catch this issue sometimes rather than not at all, but I think saying this is only a performance concern is understating it. I'd think blocking a thread on every query from the grid could pretty easily make the app completely unresponsive for other users.

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs
Comment (halter73) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs:318]: Is now the time to make this configurable?

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Infrastructure/AsyncQueryExecutorSupplier.cs
Comment (SteveSandersonMS) [reply src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Infrastructure/AsyncQueryExecutorSupplier.cs:28]: The main use cases are [A] working with in-memory data (someArray.AsQueryable, when you already got the data from somewhere - this is especially common on WebAssembly) or [B] Entity Framework or [C] remote API calls. None of these cases will result in thread-blocking I/O, so in mainstream scenarios all is well.

As for the uncommon case of an IQueryable that isn't from EF and does thread-blocking I/O, that's not good but is it really our business to actually stop people from doing it? How would we even know it does thread-blocking I/O? Should other parts of the product detect if you're using thread-blocking I/O from an IQueryable and stop it? I'm not sure there's much we could or should do about this, but am definitely open to suggestions!

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs
Comment (Nick-Stanton) [reply src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs:318]: I don't think so. I feel like there's a few ways to pursue this particular TODO that would require some time to figure out. I think I'll take the small handful of remaining TODOs in here and make small tasks on GitHub that community members could grab - or us depending on capacity.

**Comment by Nick-Stanton** [comment]
@dotnet/aspnet-blazor-eng @dotnet/aspnet-build This should be ready for final review. Failures are in the quarantined pipeline. The last commit includes everything from API review with the exception of a couple findings in an email thread. I will also include them below:

1.	Adding Attributes dictionary to the grid – historically, this is accomplished relatively easily during a BuildRenderTree() override that many of the components we (Blazor) provide have built-in. However, the current implementation doesn’t allow for this override due to the complexity of the QuickGrid component. Because of this, I could not find an easy way to add an unknown number of attributes to this component.
2.	We discussed obsoleting ISortBuilderColumn and consolidating as much as possible into ColumnBase. While trying to implement this I found that it became quite convoluted. In implementation, much of the sorting logic thrives on using the interface as a first-pass filter for sorting compatibility (see GridItemsProviderRequest where the presence of the interface simplifies a lot) and I found it awkward trying to implement a parameter that should be hidden in a class that inherits it (PropertyColumn implementing SortBy through ColumnBase). Also, if we’re moving SortBy to ColumnBase then the value of TemplateColumn is diminished greatly.

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter/src/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter.csproj
Comment (dougbu) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter/src/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter.csproj:5]: Please remove this line. It's the default for all projects targeting only `net8.0`. See https://github.com/dotnet/aspnetcore/blob/main/eng/targets/CSharp.Common.targets#L50-L53

Same for the other new project.

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter/src/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter.csproj
Comment (dougbu) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter/src/Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter.csproj:7]: If this project will produce a shipping package, it should probably have `$(Description)` and `$(GenerateDocumentationFile)` properties. Also, check the src/Components/Directory.Build.props file to confirm the `$(PackageTags)` property there is sufficient.

Same for the other new project except it at least has a `$(Description)` property.

If this will produce a transport package, add `<IsShippingPackage>false</IsShippingPackage>` instead. I doubt this should be non-packable (because I don't see anything wrapping the built content) but `<IsPackable>false</IsPackable>` would flip that bit.

**Review by dougbu — COMMENTED** [review]
@dotnet/aspnet-build only owns a couple of files in this PR. I made a couple of comments there but am not approving because I ignored almost everything else.

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/GridItemsProvider.cs
Comment (Nick-Stanton) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/GridItemsProvider.cs:11]: While _technically_ not correct, the compiler would not eat the original line. This may be due to the existence of a struct with a generic parameter and a static class (that just grabs instances of the struct) with the same name. Regardless, the actual output of this is reduced to `ValueTask<TResult>` anyways, so this clarification isn't any more useful.

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/GridItemsProviderRequest.cs
Comment (halter73) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/GridItemsProviderRequest.cs:65]: Nit:

```suggestion
    public IQueryable<TGridItem> ApplySorting(IQueryable<TGridItem> source) =>
        SortByColumn?.SortBy?.Apply(source, SortByAscending) ?? source;
```

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/GridItemsProviderRequest.cs
Comment (halter73) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/GridItemsProviderRequest.cs:75]:
```suggestion
    public IReadOnlyCollection<SortedProperty> GetSortByProperties() =>
        SortByColumn?.SortBy?.ToPropertyList(SortByAscending) ?? Array.Empty<SortedProperty>();
```

**Review by halter73 — APPROVED** [review]

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Columns/Align.cs
Comment (MackinnonBuck) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Columns/Align.cs:32]: Super duper nit: Could we add a '.' at the end of each comment that doesn't have one?

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Columns/SortedProperty.cs
Comment (MackinnonBuck) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/Columns/SortedProperty.cs:7]:
```suggestion
/// Holds the name of a property and the direction to sort by.
```
I think "ValueTuple" implies `System.ValueTuple`, but this is just a normal struct.

**Review by MackinnonBuck — APPROVED** [review]

**Comment by Nick-Stanton** [comment]
Blocked on timeouts in the Helix pipeline

**Comment by Nick-Stanton** [comment]
@SteveSandersonMS has offered to bring this across the finish line now that I'm in Azure Core. Here's what's remaining:

1. Everything is good to merge once the Helix timeline _doesn't_ time out. Please make sure that Steve is listed as a co-author in the merge commit 😄 
2. Backport to .NET 8 preview 2. This is very low risk because these projects are being consumed as NuGet packages.
3. Ping Artak

CC: @mkArtakMSFT

**Comment by SteveSandersonMS** [comment]
Thanks very much @Nick-Stanton! It's merged, the backport PR is [here](https://github.com/dotnet/aspnetcore/pull/46975), and @mkArtakMSFT has been pinged.

**Comment by Nick-Stanton** [comment]
> > This looks good to me!
> > Should we consider sealing some (most?) of these public types? I don't see why someone would need to derive from `GridSort<TGridItem>`, or `Defer`, for example.
> 
> Because you can't envision a need to derive from it, it shouldn't be open to it? That seems flawed. Why don't we seal things that we don't want inherited for a specific reason?

@sjd2021 in general, types that aren't intended to be extensible are sealed to provide [potential startup and size optimizations](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/#peanut-butter). While this guidance generally applies to private types within .NET, limitations in the Razor compiler prevented some of these types from being non-public. QuickGrid by design has multiple points of extension for developers that want to build on top of it without a custom fork of the runtime, but just not at the layer of types like `GridSort<TGridItem>` and `Defer`. If you disagree with this process, then I encourage you to [open an issue](https://github.com/dotnet/aspnetcore/issues) so that the Blazor team can discuss further with you - I recently moved to a different team within Microsoft so my understanding of these decisions may be outdated.

**Pull request merged**
